_posts/2020-06-28-openfaas-loki-into.md

```diff
@@ -0,0 +1,235 @@
+2020-06-27T16:37:17Z 2020/06/27 16:37:17 Forking fprocess.
+2020-06-27T16:37:17Z 2020/06/27 16:37:17 Wrote 119 Bytes - Duration: 0.130287 seconds
+```
+
+We will see a slightly different format when we switch to Loki later.
+
+### Install Loki and Grafana
+
+With `arkade`, we need this single line to install both Grafana _and_ Loki:
+
+```bash
+arkade install loki --grafana
+```
+
+Wati for Loki and Grafana to be ready:
+
+```bash
+kubectl rollout status deployment loki-stack-grafana
+```
+
+Now get your admin password and forward the Grafana port
+```bash
+kubectl get secret loki-stack-grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo
+kubectl port-forward service/loki-stack-grafana 3000:80 &
+```
+
+Then open `http://localhost:3000` and login using username `admin`.
+
+The cool thing about Loki is that it will automatically start collecting logs from _all_ of the Pods it can sees. Importantly, this means you can see the logs for all of the OpenFaaS components, like the Gateway.
+
+This [link](http://localhost:3000/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Loki%22,%7B%22expr%22:%22%7Bapp%3D%5C%22gateway%5C%22%7D%22%7D,%7B%22mode%22:%22Logs%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D) will show you the logs (over the last hour) for the Gateway we just installed!
+
+Loki, when installed in Kubernetes, will automatically create a search index for each of the lables on the Pods it sees. Important, this means the Loki query for the logs of an OpenFaaS function looks like `{faas_function="<name>"}` because OpenFaaS always adds the function name as the label `faas_function`.  This [link](http://localhost:3000/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Loki%22,%7B%22expr%22:%22%7Bfaas_function%3D%5C%22cows%5C%22%7D%22%7D,%7B%22mode%22:%22Logs%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D) will show the logs for the Cows function we installed and invoked.
+
+### Installing `openfaas-loki`
+
+To combine OpenFaaS and Loki, we need to install `openfaas-loki`:
+
+```bash
+arkade install openfaas-loki
+```
+
+We then need to configure the OpenFaaS installation, we can do this with `arkade` using the `--set` flag
+
+```bash
+arkade install openfaas --basic-auth=false --set gateway.logsProviderURL=http://openfaas-loki.default:9191/
+```
+
+## Testing the logs
+
+### Recreate the port-forward
+Earlier, while we were installing and verifying the system, we ran this command
+
+```bash
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+Unfortunately, that connection is no longer valid, because we updated the OpenFaaS installation to configure the logs provider.  We need to stop the current `port-forward` command and then start a new one:
+
+```bash
+pkill kubectl
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+### Invoke the function and check the logs
+Now we can use the `faas-cli` to invoke and see the logs, this time powered by Loki.
+
+```bash
+faas-cli invoke cows <<< ""
```

Comment:
The ` <<< ""` syntax may be unclear to some users (I learned it from you using it actually) - how about `echo | ` or saying that you will send an empty string as the input?